paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/source/operator/MonitorSource.java

```diff
@@ -0,0 +1,304 @@
+            nextSnapshots.clear();
+        }
+
+        @Override
+        public List<SimpleSourceSplit> snapshotState(long checkpointId) {
+            this.checkpointState.clear();
+            Long nextSnapshot = this.scan.checkpoint();
+            if (nextSnapshot != null) {
+                this.checkpointState.add(nextSnapshot);
+                this.nextSnapshotPerCheckpoint.put(checkpointId, nextSnapshot);
+            }
+
+            List<Tuple2<Long, Long>> nextSnapshots = new ArrayList<>();
+            this.nextSnapshotPerCheckpoint.forEach((k, v) -> nextSnapshots.add(new Tuple2<>(k, v)));
+            this.nextSnapshotState.update(nextSnapshots);
+
+            if (LOG.isDebugEnabled()) {
+                LOG.debug("{} checkpoint {}.", getClass().getSimpleName(), nextSnapshot);
+            }
+
+            List<SimpleSourceSplit> results = new ArrayList<>();
+            results.addAll(checkpointState.snapshotState());
+            results.addAll(nextSnapshotState.snapshotState());
+            return results;
+        }
+
+        @Override
+        public void addSplits(List<SimpleSourceSplit> list) {
+            LOG.info("Restoring state for the {}.", getClass().getSimpleName());
+            checkpointState.restoreState(list);
+            nextSnapshotState.restoreState(list);
+
+            List<Long> retrievedStates = checkpointState.get();
+
+            // given that the parallelism of the source is 1, we can only have 1 retrieved items.
+            Preconditions.checkArgument(
+                    retrievedStates.size() <= 1,
+                    getClass().getSimpleName() + " retrieved invalid state.");
+
+            if (retrievedStates.size() == 1) {
+                this.scan.restore(retrievedStates.get(0));
+            }
+
+            for (Tuple2<Long, Long> tuple2 : nextSnapshotState.get()) {
+                nextSnapshotPerCheckpoint.put(tuple2.f0, tuple2.f1);
+            }
+        }
+
+        @Override
+        public InputStatus pollNext(ReaderOutput<Split> readerOutput) throws Exception {
+            boolean isEmpty;
+            try {
+                List<Split> splits = scan.plan().splits();
+                isEmpty = splits.isEmpty();
+                splits.forEach(readerOutput::collect);
+
+                if (emitSnapshotWatermark) {
+                    Long watermark = scan.watermark();
+                    if (watermark != null) {
+                        readerOutput.emitWatermark(new Watermark(watermark));
+                    }
+                }
+            } catch (EndOfScanException esf) {
+                LOG.info("Catching EndOfStreamException, the stream is finished.");
+                return InputStatus.END_OF_INPUT;
+            }
+
+            if (isEmpty) {
+                Thread.sleep(monitorInterval);
+            }
+            return InputStatus.MORE_AVAILABLE;
+        }
+    }
+
+    private static class SplitState<T> {
```

Comment:
Make it a public utility class. You also store state values into splits in tests, for example `FiniteTestSource`.